### PR TITLE
LIVY-280. Livy session leakage issue when app submission timeout

### PR DESCRIPTION
**Pull request by zjffdu**
* Add a new daemon thread to kill leaked livy session regularly.  
* Change the default look-up timeout to 60s.

### CODE REVIEW

**Comment by jerryshao** [comment]
Did it fixe by this [commit](https://github.com/cloudera/livy/commit/e82780b96affc4d1cf8a1d08d41ac98844c2af62)?

**Comment by zjffdu** [comment]
No. LIVY-251 resolve the session leakage when yarn app is launched but stuck in ACCEPTED state. While this PR fix the issue when even AM is not started due to resource limitation.

**Comment by jerryshao** [comment]
I think LIVY-251 can also kill the stuck application in which AM cannot be started due to resource limitation. I tried before.

**Comment by zjffdu** [comment]
I don't see the yarn app is killed. You can set `livy.server.yarn.app-lookup-timeout` to `5s` to reproduce the issue.

**Comment by zjffdu** [comment]
@tc0312 @linchan-ms Please help review.

**Comment by tc0312** [comment]
From our testing, LIVY-251, works also when AM is not started. Do you have a reliable way to repro the leakage you are seeing? May be there's another edge case.

**Comment by zjffdu** [comment]
@tc0312 I can reproduce this issue even with LIVY-251. You can adjust livy.server.yarn.app-lookup-timeout and livy.server.yarn.poll-interval to make the app submission fails. And I don't see LIVY-251 handle the situation I mentioned here. 

This is the error what I see and the yarn app is still running.

```
{
  "from": 0,
  "total": 1,
  "sessions": [
    {
      "id": 0,
      "appId": null,
      "owner": null,
      "proxyUser": null,
      "state": "dead",
      "kind": "spark",
      "appInfo": {
        "driverLogUrl": null,
        "sparkUiUrl": null
      },
      "log": [
        "java.lang.Exception: No YARN application is tagged with livy-session-0-zyl0mqvy.",
        "com.cloudera.livy.utils.SparkYarnApp.com$cloudera$livy$utils$SparkYarnApp$$getAppIdFromTag(SparkYarnApp.scala:131) com.cloudera.livy.utils.SparkYarnApp$$anonfun$1$$anonfun$4.apply(SparkYarnApp.scala:195) com.cloudera.livy.utils.SparkYarnApp$$anonfun$1$$anonfun$4.apply(SparkYarnApp.scala:192) scala.Option.getOrElse(Option.scala:120) com.cloudera.livy.utils.SparkYarnApp$$anonfun$1.apply$mcV$sp(SparkYarnApp.scala:192) com.cloudera.livy.Utils$$anon$1.run(Utils.scala:95)"
      ]
    }
  ]
}
```

**Comment by jerryshao** [comment]
@zjffdu , I think I could reproduce your issue, this is a different issue compared to LIVY-251. In your case the problem is mainly due to the timeout of this API `getAppIdFromTag`, looks like spark hasn't yet request application id from RM when this API was timed-out.

File: server/src/main/scala/com/cloudera/livy/utils/SparkYarnApp.scala [hunk 1]
```diff
@@ -50,6 +58,47 @@ object SparkYarnApp extends Logging {
...
+            val entry = iter.next()
+            yarnClient.getApplications(appType).asScala
+              .find(_.getApplicationTags.contains(entry.getKey))
+              .foreach(e => {
+                info(s"Kill leakage app ${e.getApplicationId}")
+                yarnClient.killApplication(e.getApplicationId)
+                iter.remove()
+                isRemoved = true
+              })
+            if (!isRemoved) {
```
Comment (jerryshao) [inline server/src/main/scala/com/cloudera/livy/utils/SparkYarnApp.scala]: When will we go into this branch?

Comment (zjffdu) [inline server/src/main/scala/com/cloudera/livy/utils/SparkYarnApp.scala]: It happens when AM still doesn't start.

File: server/src/main/scala/com/cloudera/livy/utils/SparkYarnApp.scala [hunk 2]
```diff
@@ -50,6 +58,47 @@ object SparkYarnApp extends Logging {
 
   private def getYarnPollInterval(livyConf: LivyConf): FiniteDuration =
     livyConf.getTimeAsMs(LivyConf.YARN_POLL_INTERVAL) milliseconds
+
+  private val appType = Set("SPARK").asJava
+
+  private val leakagedAppTags = new java.util.concurrent.ConcurrentHashMap[String, Long]()
```
Comment (tc0312) [inline server/src/main/scala/com/cloudera/livy/utils/SparkYarnApp.scala]: leakagedAppTags is not persisted across restart. We should persist it to the state store.

File: server/src/main/scala/com/cloudera/livy/utils/SparkYarnApp.scala [hunk 3]
```diff
@@ -50,6 +58,47 @@ object SparkYarnApp extends Logging {
...
+
+  private val leakagedAppGCThread = new Thread() {
+    override def run(): Unit = {
+      while (true) {
+        if (!leakagedAppTags.isEmpty) {
+          // kill the app if found it and remove it if exceeding a threashold
+          val iter = leakagedAppTags.entrySet().iterator()
+          var isRemoved = false
+          val now = System.currentTimeMillis()
+          while(iter.hasNext) {
```
Comment (tc0312) [inline server/src/main/scala/com/cloudera/livy/utils/SparkYarnApp.scala]: Is it possible to batch the getApplications() calls? I think we can do it by using RMClient directly without going thru YarnClient.

Comment (zjffdu) [inline server/src/main/scala/com/cloudera/livy/utils/SparkYarnApp.scala]: I move `yarnClient.getApplications` to the outside of while loop so that we only call it one time.  LIVY-214 is only for hadoop-2.9, we can do it in another PR.

File: server/src/main/scala/com/cloudera/livy/utils/SparkYarnApp.scala (hunk 2 above)
Comment (zjffdu) [inline server/src/main/scala/com/cloudera/livy/utils/SparkYarnApp.scala]: I tried it, but it looks like it is a little complicated than I expect. Because it may affect the code of batch session recovering. The remaining issue is that the leaked session can not be cleanup after recovery, I think this is not critical, so we can do it in a follow up ticket, what do you think ?

File: server/src/main/scala/com/cloudera/livy/utils/SparkYarnApp.scala [hunk 4]
```diff
@@ -50,6 +59,47 @@ object SparkYarnApp extends Logging {
...
+        if (!leakedAppTags.isEmpty) {
+          // kill the app if found it and remove it if exceeding a threashold
+          val iter = leakedAppTags.entrySet().iterator()
+          var isRemoved = false
+          val now = System.currentTimeMillis()
+          val apps = yarnClient.getApplications(appType).asScala
+          while(iter.hasNext) {
+            val entry = iter.next()
+            apps.find(_.getApplicationTags.contains(entry.getKey))
+              .foreach(e => {
```
Comment (jerryshao) [inline server/src/main/scala/com/cloudera/livy/utils/SparkYarnApp.scala]: Mind if you change to

```
.foreach { e =>
...
...
}
```

File: server/src/main/scala/com/cloudera/livy/server/LivyServer.scala [hunk 5]
```diff
@@ -90,6 +90,7 @@ class LivyServer extends Logging {
 
     // Initialize YarnClient ASAP to save time.
     if (livyConf.isRunningOnYarn()) {
+      SparkYarnApp.init(livyConf)
```
Comment (jerryshao) [inline server/src/main/scala/com/cloudera/livy/server/LivyServer.scala]: It is not necessary to have a `init(...)` function to create and start the GC thread. We could directly initialize the values inside the `SparkYarnApp`. This values will be initialized when `SparkYarnApp` class is loaded, which is here calling `SparkYarnApp.yarnClient`.

File: server/src/main/scala/com/cloudera/livy/utils/SparkYarnApp.scala (hunk 2 above)
Comment (jerryshao) [inline server/src/main/scala/com/cloudera/livy/utils/SparkYarnApp.scala]: Can you please create a JIRA to track this issue.

File: server/src/main/scala/com/cloudera/livy/utils/SparkYarnApp.scala (hunk 4 above)
Comment (zjffdu) [inline server/src/main/scala/com/cloudera/livy/utils/SparkYarnApp.scala]: addressed

**Comment by zjffdu** [comment]
@tc0312 Any more comments ?

File: server/src/main/scala/com/cloudera/livy/utils/SparkYarnApp.scala (hunk 3 above)
Comment (tc0312) [inline server/src/main/scala/com/cloudera/livy/utils/SparkYarnApp.scala]: If we use RMClient directly, I think we can do it without depending on YARN 2.9.

File: conf/livy.conf [hunk 6]
```diff
@@ -75,7 +75,13 @@
 # livy.server.recovery.state-store.url =
 
 # If Livy can't find the yarn app within this time, consider it lost.
-# livy.server.yarn.app-lookup-timeout = 30s
+# livy.server.yarn.app-lookup-timeout = 60s
+# When the cluster is busy, we may fail to launch yarn app in 30s, then it would cause session
```
Comment (tc0312) [inline conf/livy.conf]: +# When the cluster is busy, we may fail to launch yarn app in app-lookup-timeout, then it would cause session

**Comment by tc0312** [comment]
Mind if you add an integration test for this?

**Comment by zjffdu** [comment]
It seems not easy to simulate the session leakage in integration test, I will try to do that in other ticket.

**Comment by zjffdu** [comment]
Merge it and create LIVY-301 as followup
